woocommerce/woocommerce-google-analytics-integration
# gtag.js and Google Analytics 4 support

### All Submissions:

* [x] *__N/A__* (Have you written new tests for your changes, as applicable?
* [x] *__N/A__* Have you successfully run tests with your changes locally?





### Changes proposed in this Pull Request:

This PR has turned out to be a bit of a whopper 🐳 . It fishes the changes proposed in #114 out of the bin to implement gtag.js, and builds on them to allow new users to get up and running with Google Analytics 4, and provide existing users with an upgrade path (should they wish to do so).

Closes #173 .

#### An overview of the changes:

<details>
<summary>Settings</summary>

  - Adds the _Use Global Site Tag_ option.
    - When selected, hides the _Enable Universal Analytics_ and Set Domain Name" options. (UA because `gtag.js` supplants `analytics.js`, and Set Domain Name because cross-domain in GA4 is [handled in the dashboard](https://support.google.com/analytics/answer/10071811?hl=en&ref_topic=9303319); in UA with gtag.js, [it's a bit trickier](https://developers.google.com/analytics/devguides/collection/gtagjs/cross-domain) so for now it's just hidden, too).
  - Updated the _Google Analytics ID_ label (now includes _Tracking_) and included the new formats in the description.
  - Updated the description for _Enable Universal Analytics_ and _Enable Enhanced eCommerce_ to indicated that either UA or Global Site Tag must be enabled to use Enhanced eCommerce.
</details>

<details>
<summary>Onboarding</summary>

  -    New installs will have the _Use Global Site Tag_ option selected, and can use either a `G-XXXXX` GA4 Measurement ID _or_ a `UA-XXXXX-X` Universal Analytics Tracking ID.
  - Updates will maintain their current settings (_Use Global Site Tag_ defaults to disabled, _Enable Universal Analytics_ set to whatever was already selected.
</details>

<details>
<summary>Developer</summary>

	- The original class that handled the analytics JS code generation was abstracted, and then two child classes were created: one for UA, the other for Gtag.
	- All Universal Analytics functionalities ported into Global Site Tag functionalities. Users _could_ just turn on Global Site Tag, but it _would_ break any custom tracking code they have (and may change some event names or attributes).
	- A quick refresh of all PHPDoc.
	- Added `gtag_enabled`, `set_domain_name`, `ga_id`, and `plugin_version` to the WC Tracker data.
   - Added the WC Developer ID for Gtag and UA.
	- JS
		- Updated show/hide/toggle logic for the settings page.
		- Removed `.click()`, deprecated in jQuery 3 (for [WordPress 5.7 compat](https://make.wordpress.org/core/2020/06/29/updating-jquery-version-shipped-with-wordpress/)).
	- Updated _Settings_ link from the plugins page (previously pointed to the Settings → Integration tab, but not to the Google Analytics sub-section).
</details>

#### And some possible To Dos (for this or following versions):</summary>
<details>
<summary>Toggle</summary>

- Conditional field descriptions based on the status of _Use Global Site Tag_ and _Enable Universal Analytics_ selections (e.g., only show the valid Tracking ID formats).
- Validate Tracking ID format depending on selected options.
- JS errors when _Enable Standard Tracking_ is not selected and _Add to cart_ is clicked from a product list page.
- Bump minimum WooCommerce to 3.0.
- Include all _Enhanced eCommerce_ interaction options in WC Tracker.
- Don't show the "Link WooCommerce and Google Analytics" notification when on the plugin settings page (the link just reloads the page).
- Documentation
	- Indicate that the plugin is compatible with Classic Analytics (ga.js), Universal Analytics (analytics.js or gtag.js), and now Google Analytics 4 (gtag.js).
	- In the Global Site Tag `view_item_list` event, send all products in one event instead of an individual event for each product in the list.
	- Make reference to the hook for using both [GA4 and UA with gtag.js](https://developers.google.com/analytics/devguides/collection/ga4/basic-tag?technology=gtagjs): 
```php
// From https://developers.google.com/analytics/devguides/collection/ga4/basic-tag?technology=gtagjs  
add_filter( 'woocommerce_gtag_snippet', function( $gtag_snippet ) {  
  return $gtag_snippet . "<script> gtag('config', 'G-GJRR7YFDR5'); </script>";  
} );
```
</details>

### How to test the changes in this Pull Request:

Needs to get a few extra eyeballs. I've done a lot of testing, and once merged, there will be a Release Candidate period and a call for testing. The most important things to confirm are:
1. Update path: stores updating to this version should maintain their settings and functionality (invisible update), but can enable Global Site Tags if they wish.
2. New installs: fresh installs should have _Use Global Site Tag_ enabled and _Enable Universal Analytics_ and _Set Domain Name_ hidden.
3. Functionality: Universal Analytics should still work the same, and both UA Tracking IDs (`UA-XXXXX-X`) and GA4 Measurement IDs (`G-XXXXX`) should report correctly when _Use Global Site Tag_ is selected.
4. Analytics Tracking: Confirm that all the possible tracking events are sent in UA and Gtag when enabled:
	- Add to cart
	- Remove from cart
	- Purchase complete
	- Checkout begun
	- Product list viewed
	- Product impression in product list
	- Product click in product list
	- Product detail viewed
	- 404 Errors
5. Confirm _Anonymize IP_, _Use Enhanced Link Attribution_, and _"Display Advertising" Support_ are configured correctly in the tracking library JS code.
5. Confirm toggling of fields on the Settings page (`/wp-admin/admin.php?page=wc-settings&tab=integration&section=google_analytics`) works as expected and the descriptions are accurate.
	- "As expected": if _Use Global Site Tag_ is selected, _Enable Universal Analytics_ and _Set Domain Name_ should be hidden. If neither _enable Universal Analytics_ nor _Use Global Site Tag_ are selected, then _Enable Enhanced eCommerce_ an its associated interactions should be hidden.


### Other information:




### Changelog entry

>   * Add - Option to use Global Site Tag and the gtag.js library (for Universal Analytics or Google Analytics 4).  
> * Add - Several new settings added to the Tracker data.  
> * Add - Developer ID for gtag.js and analytics.js.  
> * Tweak - Remove deprecated jQuery .click().  
> * Fix - Settings link in plugins table row points to plugin settings correctly.
